File: backend/models/reminder.py
from pydantic import BaseModel, Field
from typing import List, Optional, Literal
from .common import TranscriptEntry

# --- OUTCOME TYPES ---

# Represents the status of the VOICE CALL itself
CallStatus = Literal["answered", "missed", "voicemail", "busy", "failed"]

# Represents the status of the TASK (Medication/Activity)
ActionStatus = Literal[
    "completed",      # Took meds / Finished exercise
    "delayed",        # User said "Call me back in 10 mins"
    "refused",        # User explicitly said "I won't do it"
    "partially_done",  # Took 1 pill out of 2
    "ignored",        # Call was answered but user hung up or stayed silent
    "pending"         # Default state
]

class ReminderDetails(BaseModel):
    """Details about the specific task assigned to the AI"""
    reminder_id: str = Field(..., description="Unique ID from Firestore")
    category: Literal["medication", "activity", "appointment"] = Field(...)
    
    # Task specific details
    title: str = Field(..., description="e.g., 'Morning Yoga' or 'Blood Pressure Pill'")
    medication_name: Optional[str] = None
    dosage: Optional[str] = None
    activity_type: Optional[str] = Field(None, description="e.g., 'Walking', 'Stretching'")
    
    # Timing logic
    scheduled_time: str = Field(..., description="ISO timestamp for the cron job")
    is_recurring: bool = False

class ReminderCall(BaseModel):
    """The complete record of a single reminder interaction"""
    user_id: str = Field(...)
    call_id: str = Field(...)
    
    # The "What"
    reminder_details: ReminderDetails = Field(...)
    
    # The "Outcome" (The Intelligent Logic Part)
    call_status: CallStatus = Field(default="missed")
    action_status: ActionStatus = Field(default="pending")
    
    # Rescheduling logic
    is_rescheduled: bool = Field(default=False)
    new_scheduled_time: Optional[str] = Field(
        None, description="If status is 'delayed', when is the next attempt?"
    )
    
    # AI Analysis
    user_reasoning: Optional[str] = Field(
        None, description="Why did they delay? e.g., 'Not hungry yet', 'Feeling tired'"
    )
    transcript: List[TranscriptEntry] = Field(default_factory=list)
    
    # Timestamps
    started_at: str = Field(...)
    ended_at: Optional[str] = None

File: mobile/lib/models/reminder.dart
class ReminderDetails {
  final String reminderId;
  final String category;
  final String title;
  final String? medicationName;
  final String? dosage;
  final String? activityType;
  final String scheduledTime;
  final bool isRecurring;

  ReminderDetails({
    required this.reminderId,
    required this.category,
    required this.title,
    this.medicationName,
    this.dosage,
    this.activityType,
    required this.scheduledTime,
    required this.isRecurring,
  });

  factory ReminderDetails.fromJson(Map<String, dynamic> json) {
    return ReminderDetails(
      reminderId: json['reminder_id'],
      category: json['category'],
      title: json['title'],
      medicationName: json['medication_name'],
      dosage: json['dosage'],
      activityType: json['activity_type'],
      scheduledTime: json['scheduled_time'],
      isRecurring: json['is_recurring'] ?? false,
    );
  }
}

class Outcome {
  final String callStatus;
  final String actionStatus;
  final bool isRescheduled;
  final String? rescheduledFromTime;
  final String? userReasoning;
  final int? attemptsMade;

  Outcome({
    required this.callStatus,
    required this.actionStatus,
    required this.isRescheduled,
    this.rescheduledFromTime,
    this.userReasoning,
    this.attemptsMade,
  });

  factory Outcome.fromJson(Map<String, dynamic> json) {
    return Outcome(
      callStatus: json['call_status'],
      actionStatus: json['action_status'],
      isRescheduled: json['is_rescheduled'] ?? false,
      rescheduledFromTime:
          json['rescheduled_from_time'] ??
          json['rescheduled_from_time'] ??
          json['rescheduled_from_time'],
      userReasoning: json['user_reasoning'],
      attemptsMade: json['attempts_made'],
    );
  }
}

class TranscriptEntry {
  final String role;
  final String text;

  TranscriptEntry({required this.role, required this.text});

  factory TranscriptEntry.fromJson(Map<String, dynamic> json) {
    return TranscriptEntry(role: json['role'], text: json['text']);
  }

  static List<TranscriptEntry> listFromJson(List<dynamic> jsonList) {
    return jsonList.map((e) => TranscriptEntry.fromJson(e)).toList();
  }
}

class Meta {
  final String startedAt;
  final String endedAt;
  final int? callDurationSeconds;
  final String? voiceId;

  Meta({
    required this.startedAt,
    required this.endedAt,
    this.callDurationSeconds,
    this.voiceId,
  });

  factory Meta.fromJson(Map<String, dynamic> json) {
    return Meta(
      startedAt: json['started_at'],
      endedAt: json['ended_at'],
      callDurationSeconds: json['call_duration_seconds'],
      voiceId: json['voice_id'],
    );
  }
}

class Reminder {
  final String userId;
  final String callId;
  final ReminderDetails reminderDetails;
  final Outcome outcome;
  final List<TranscriptEntry> transcript;
  final Meta meta;

  Reminder({
    required this.userId,
    required this.callId,
    required this.reminderDetails,
    required this.outcome,
    required this.transcript,
    required this.meta,
  });

  factory Reminder.fromJson(Map<String, dynamic> json) {
    return Reminder(
      userId: json['user_id'],
      callId: json['call_id'],
      reminderDetails: ReminderDetails.fromJson(json['reminder_details']),
      outcome: Outcome.fromJson(json['outcome']),
      transcript: TranscriptEntry.listFromJson(json['transcript'] ?? []),
      meta: Meta.fromJson(json['meta']),
    );
  }

  static List<Reminder> listFromJson(List<dynamic> jsonList) {
    return jsonList.map((e) => Reminder.fromJson(e)).toList();
  }
}
